docs/proposals/trimming-friendly-registrar-proposal.md

```diff
@@ -0,0 +1,483 @@
+      var managedHandle = Messaging.IntPtr_objc_msgSend(handle, createManagedInstance);
+      return GCHandle.FromIntPtr(managedHandle).Target as T;
+    }
+
+    // fallback for C-style structs and protocols
+    var fallbackSelector = Selector.GetHandle ($"createManagedInstance_{MangleName (typeof (T))}:");
+    var canCreateManagedInstanceUsingFallback = class_respondsToSelector (dotnetClass.Handle, fallbackSelector) == 1;
+    if (canCreateManagedInstanceUsingFallback) {
+      var managedHandle = Messaging.IntPtr_objc_msgSend(dotnetClass.Handle, fallbackSelector, handle);
+      return GCHandle.FromIntPtr(managedHandle).Target as T;
+    }
+
+    return null;
+  }
+}
+```
+
+There is a similar solution for the inverse direction in which we want to create a new instance of a native object
+for an existing .NET object (which boils down to the implementation of `Class.FindClass(Type type, out bool isCustomType)`):
+
+```c#
+[Register("MyNativeClass")]
+public class MyClass : NSObject {}
+```
+```obj-c
+@implementation __dotnet (MyNativeClass)
++(id) getNativeClass_MyNamespace_MyClass: (BOOL*) isCustomType {
+  *isCustomType = YES;
+  return [MyNativeClass class];
+}
+@end
+```
+```c#
+partial class Class {
+  private static IntPtr s_dotnetTypesClass = Class.GetHandle("__dotnet");
+
+  public unsafe static IntPtr FindClass(Type type, out bool isCustomType) {
+    byte _isCustomType = 0;
+    var selector = Selector.GetHandle($"getNativeClass_{MangleName(type)}:");
+    var classHandle = Messaging.IntPtr_objc_msgSend_ref_byte(s_dotnetTypesClass, selector, &_isCustomType);
+    isCustomType = _isCustomType == 1;
+    return classHandle;
+  }
+}
+```
+
+## Dependencies between native and managed code - considerations for guiding IL Compiler
+
+The registrar allows native Objective-C code to create instances of managed classes and call methods on them. In fact,
+some managed objects might only be instantiated from Objective-C and some their methods might only ever be called from
+Objective-C. In these cases, we need to guide the trimmer to correctly preserve the code we need. Some of the scenarios
+we need to express are these:
+
+- "Keep this class and all its members. It is only ever created from Objective-C and only called from Objective-C."
+- "Keep all the members of this class if this class survives trimming. Some of the members will only be called through UCOs."
+
+In Xamarin, it was a common practice to use `[Preserve(AllMembers = true)]`  and `[Preserve(Conditional = true)]`
+attributes to express these scenarios. In NativeAOT, we can "strategically" place `[DynamicDependency]` attributes
+on module initializers or class and instance constructors to emulate similar behaviour today.
+
+Unfortunately, `DynamicDependency` implies use of reflection. Therefore, the IL Compiler emits unnecessary metadata.
+Ideally, we would have some additional attribute to express "static" or "implicit" dependency.  It would also help if we
```

Comment:
I'm confused by this - if the member in question is accessed via direct reference, then there should be no need for `DynamicDependency`. If it's accessed via reflection, then we need the reflection metadata so `DynamicDependency` is correct. Is there some other way the member can be accessed?

Reply:
I see. I think I'll drop the part about reflection completely.

The reason we need those `DynamicDependency` attributes is to emulate the `Preserve` attributes (as described in the next section). We're only using direct references, no reflection at all. `DynamicDependency` seems like an overkill but there isn't any other similar tool AFAIK.

Reply:
We can maybe discuss this offline - I would be interested in the particulars, since if there are direct references, then there should be no need for `DynamicDependency`, but this is a very special scenario, so it's likely I don't get some important detail.

Reply:
Yes, definitely. It's also quite possible that there is an obvious solution here that I am missing 🤷‍♂️

Reply:
From our offline discussion:
There is actually no hard *managed code* reference to the UCOs in question - they only have callers from Objective-C. That's why we need `DynamicDependency` on them to preserve them through the managed trimming process. The platform may be able to remove the unused ones (if we don't add exports for them), but any additional managed world dependencies from them would still be kept around.